### ISSUE TITLE
[Parquet] Encoding configuration should be easier and more automated

### ISSUE THREAD

**Issue by pitrou** [apache/arrow#38441]
### Describe the enhancement requested

Right now, configuring good encoding values is difficult for users. There is nothing to help them make those decisions, and the defaults are a bit simplistic (try RLE_DICTIONARY then fall back on PLAIN, IIUC). If they want to override encodings, they have to do so on a column-by-column basis (which probably becomes very cumbersome if there hundreds of columns).

Ideally, there should be a way for users to get an automatic selection of encodings, based on their data or at least their data types (and also the selected Parquet version), that provides a good compromise between disk footprint and decoding speed.

(in Python, think `pq.write_table(..., column_encoding="auto")`)

Perhaps it would be also nice for users to pass per-datatype preferences, rather than per-column.

### Component(s)

C++, Parquet

**Comment by pitrou** [comment]
cc @jorisvandenbossche @mapleFU @wgtmac

**Comment by pitrou** [comment]
Also @mrocklin, perhaps you have specific API suggestions here.

**Comment by jorisvandenbossche** [comment]
At the very least, we should also improve the documention around the different encoding options. Currently we just have this table listing them: https://arrow.apache.org/docs/dev/cpp/parquet.html#encodings (and this brief section in the Python user guide: https://arrow.apache.org/docs/dev/python/parquet.html#compression-encoding-and-file-compatibility) 
For example mention for which data types they are relevant (eg BYTE_STREAM_SPLIT is only for floats, DELTA_BINARY_PACKED for integers), or if they are useful for specific characteristcs (eg many repeated values, sorted data, ..), etc.

**Comment by mapleFU** [comment]
https://www.cs.cit.tum.de/fileadmin/w00cfj/dis/papers/btrblocks.pdf

Remember here is a policy here. Though I think this paper has something weird, we can be able to using it.

<img width="693" alt="0AFBE093-828B-46B9-81DD-8C3FD02A64FF" src="https://github.com/apache/arrow/assets/24351052/fd3ce6ee-0033-416f-8280-ba38e90139a8">

However, hacking the encoder would be tricky...

**Comment by wgtmac** [comment]
> https://www.cs.cit.tum.de/fileadmin/w00cfj/dis/papers/btrblocks.pdf

That paper looks a little bit academic to me as it would significantly affect the write time. In many common cases, a simple sampling process (e.g. on first N input values if not skewed) would help determine a proper encoding.

**Comment by pitrou** [comment]
Yes, and perhaps the encoding decisions can be remembered for the entire row batch (or the entire file?).

**Comment by mapleFU** [comment]
Another problem is that, for DataPageV1, we need to both considering compression and encoding(DataPage V2 can choose not to compression if compression ratio is low). This make adaptive encoding more complex.

**Comment by pitrou** [comment]
We can decide not to support DataPageV2 for this feature if it makes things easier.

**Comment by mapleFU** [comment]
Yeah, I mean, usally user would pickup a compression. For datapagev1, it doesn't support a `is_compression` flag. So, all the page should be compressed ( whenever the compression ratio is)

The funny thing is that, if we meet some data, and have `PLAIN` `DELTA` as candidate. DELTA might be better if not compression, but considering compression, PLAIN is better.

Sometimes, the compression size even grows larger for some input. DataPageV2 can reject to compress for these types, but DataPageV1 cannot.

**Comment by wgtmac** [comment]
I think it is better to decouple encoding and compression, though some compression codecs may favor different encoding schemes when data pattern varies. So it would be easier if we just focus on "best" encoding that we can decide from sample data regardless of follow-up compression. Then probably we don't have to bother with DataPage version at the moment.

**Comment by pitrou** [comment]
Well, for example, the only point of BYTE_STREAM_SPLIT is that the encoded output compresses better. But if you don't compress, it's exactly the same size as the input.

So I'm afraid the strategy should take compression into account _somehow_, if some kind of sampling is used.

That said, the strategy can also not use sampling at all, and instead rely on simple heuristics (which ones?).

All this probably means is that:
* we should offer a flexible API for users to implement their own strategy
* we should also offer a decent default strategy for the majority of users

I can draft an API when I find the time, which would be in a couple weeks probably (feel free to ping me :-)).

**Comment by mapleFU** [comment]
How could you sampling and doing this?🤔 I've implement a similiar one but I must say it isn't suitable for all cases and the interface would be tricky(for example, handling `dict` is tricky). Also, when `NULL FIRST` is setting, we cannot sampling any data.

**Comment by mapleFU** [comment]
Also, would it related to: https://github.com/apache/arrow/issues/33885 ?

**Comment by pitrou** [comment]
Slighty related, but also independent IMHO.

**Comment by wgtmac** [comment]
Parquet is usually used as a low-level library in the database or data warehouse, which has a better knowledge of the actual data pattern. They can decide the best combination of encoding and compression for any column based on the data already stored. Should we provide a simple tool or something to evaluate each page of every column from an input parquet file and decides the best encoding & compression for them?

**Comment by pitrou** [comment]
> Should we provide a simple tool or something to evaluate each page of every column from an input parquet file and decides the best encoding & compression for them?

Such a tool can easily be developed as a separate project. Here we're talking about allowing users to get better defaults out of Parquet C++.

**Comment by thisisnic** [comment]
This issue has been marked as stale because it has had no activity in the past 2 years. Please remove the stale label or comment below, or this issue will be closed in 14 days.

**Comment by daniel-x** [comment]
I'm interested to work on this issue, as I see high potential in it for reduced file sizes, which is a strong reason for choosing parquet as a file format for long term storage - and we would improve io speed at the same time. I personally work with data where slightly more detailed encoding defaults would reduce my zstd compressed file sizes by about 20%. Currently, specifying encoding per-column through the Python API requires a fair bit of manual configuration, which can be a barrier for new users.

Here is how I think about scoping a first step:
- I suggest targeting simple, low-risk changes that harvest large gains.
- We can set defaults for encoding per column based on
  - datatype of the column
  - subsequent compression
  - version of the parquet file (for compatibility)     # can be omitted and simplified by choosing always-compatible defaults
  - (no sampling of column content in this first step, because this is where it gets tricky)

I believe this will address many of the common cases.

Technically, encoding can be decoupled from subsequent compression, but it can cause larger file sizes and worse io speed, as demonstrated by the benchmarks in issue #49715. Hence, I think encoding should take compression into account in order to get best results, even though it increases code complexity.

I'd be happy to put together an initial PR if this direction sounds reasonable. Happy to hear if there are considerations I'm missing.

**Comment by daniel-x** [comment]
Please let me know if there is interest by other developers in me implementing something and/or if I should use another medium for the discussion, like the mailing list or something else.

**Comment by kou** [comment]
Sorry for no response in a month.

Could you also try `dev@arrow.apache.org` https://arrow.apache.org/community/#mailing-lists for visibility?
